source/secure_tunneling.c

```diff
@@ -529,142 +452,1400 @@ static int s_init_data_tunnel_pair(
+
+    s_aws_secure_tunnel_operational_state_reset(
+        secure_tunnel, AWS_ERROR_IOTDEVICE_SECURE_TUNNELING_USER_REQUESTED_STOP);
+
+    /* Stop works as a complete session wipe, and so the next time we connect, we want it to be clean */
+    s_reset_secure_tunnel(secure_tunnel);
+
+    if (secure_tunnel->config->on_stopped) {
+        secure_tunnel->config->on_stopped(secure_tunnel->config->user_data);
+    }
+}
+
+static void s_change_current_state_to_connecting(struct aws_secure_tunnel *secure_tunnel) {
+    AWS_ASSERT(
+        secure_tunnel->current_state == AWS_STS_STOPPED || secure_tunnel->current_state == AWS_STS_PENDING_RECONNECT);
+
+    secure_tunnel->current_state = AWS_STS_CONNECTING;
+
+    int result = s_websocket_connect(secure_tunnel);
+
+    if (result) {
+        int error_code = aws_last_error();
+        AWS_LOGF_INFO(
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: failed to kick off connection with error %d(%s)",
+            (void *)secure_tunnel,
+            error_code,
+            aws_error_debug_str(error_code));
+
+        s_change_current_state(secure_tunnel, AWS_STS_PENDING_RECONNECT);
+    }
+}
+
+static void s_change_current_state_to_connected(struct aws_secure_tunnel *secure_tunnel) {
+    AWS_FATAL_ASSERT(secure_tunnel->current_state == AWS_STS_CONNECTING);
+
+    secure_tunnel->current_state = AWS_STS_CONNECTED;
+    secure_tunnel->pending_write_completion = false;
+    secure_tunnel->reconnect_count = 0;
+
+    /*
+     * TODO Any rejoin logic can be implemented here. Secure Tunnel does not handle any rejoin state.
+     * We may opt to send disconnects to existing non-zero stream IDs to notify that the server has reconnected.
+     */
+
+    s_reset_ping(secure_tunnel);
+}
+
+static void s_change_current_state_to_clean_disconnect(struct aws_secure_tunnel *secure_tunnel) {
+    AWS_FATAL_ASSERT(secure_tunnel->current_state == AWS_STS_CONNECTED);
+
+    secure_tunnel->current_state = AWS_STS_CLEAN_DISCONNECT;
+}
+
+static void s_change_current_state_to_websocket_shutdown(struct aws_secure_tunnel *secure_tunnel) {
+    enum aws_secure_tunnel_state current_state = secure_tunnel->current_state;
+    AWS_FATAL_ASSERT(
+        current_state == AWS_STS_CONNECTING || current_state == AWS_STS_CONNECTED ||
+        current_state == AWS_STS_CLEAN_DISCONNECT);
+
+    if (secure_tunnel->websocket) {
+        aws_websocket_close(secure_tunnel->websocket, false);
+    } else {
+        s_on_websocket_shutdown(secure_tunnel->websocket, AWS_ERROR_UNKNOWN, secure_tunnel);
+    }
+
+    secure_tunnel->current_state = AWS_STS_WEBSOCKET_SHUTDOWN;
+}
+
+static void s_update_reconnect_delay_for_pending_reconnect(struct aws_secure_tunnel *secure_tunnel) {
+
+    uint64_t delay_ms = MIN_RECONNECT_DELAY_MS;
+    for (int i = 0; i < (int)secure_tunnel->reconnect_count; ++i) {
+        delay_ms = delay_ms * 2;
+    }
+
+    delay_ms = aws_min_u64(delay_ms, MAX_RECONNECT_DELAY_MS);
```

Comment:
I feel like the single destination device connecting to a single secure tunnel connection expectation makes jitter support overkill for secure tunnels since there won't be any realistic chance of a ton of these things synchronizing connection attempts due to lack of jitter. Open to adding if you feel differently or I'm missing a use case.

Reply:
The issue is when the service goes down and every connection drops simultaneously

Reply:
Ah, right. That makes sense. I'll add this to a stretch goal for V3!